src/ReportPortal.Shared/Internal/Delegating/RequestExecutionThrottler.cs

```diff
@@ -35,28 +35,23 @@ public RequestExecutionThrottler(int maxConcurrentRequests)
         /// <inheritdoc/>
         public async Task ReserveAsync()
         {
-            TraceLogger.Verbose($"Awaiting free executor. Currently available: {_concurrentAwaiter.CurrentCount}");
+            TraceLogger.Verbose($"Awaiting free executor. Available: {_concurrentAwaiter.CurrentCount}, demand: {_waitingThreads}");
 
-            await _concurrentAwaiter.WaitAsync().ConfigureAwait(false);
+            Interlocked.Increment(ref _waitingThreads);
 
-            TraceLogger.Verbose($"Executer is reserved. Available executers after reservation: {_concurrentAwaiter.CurrentCount}");
+            await _concurrentAwaiter.WaitAsync().ConfigureAwait(false);
 
-            // track maximum level of actual concurrent requests
-            lock (_concurrentAwaiter)
-            {
-                var maximumConcurrentCountRequested = MaxCapacity - _concurrentAwaiter.CurrentCount;
-                if (maximumConcurrentCountRequested > _maximumConcurrentCountRequested)
-                {
-                    _maximumConcurrentCountRequested = maximumConcurrentCountRequested;
-                    TraceLogger.Verbose($"Maximum level of actual concurrent requests has been changed to {_maximumConcurrentCountRequested}");
-                }
-            }
+            TraceLogger.Verbose($"Executor is reserved. Available: {_concurrentAwaiter.CurrentCount}");
         }
 
         /// <inheritdoc/>
         public void Release()
         {
-            _concurrentAwaiter.Release();
+            var previousCount = _concurrentAwaiter.Release();
+
+            Interlocked.Decrement(ref _waitingThreads);
```

Comment:
Should this be up within the `ReserveAsync` method instead?

It's no longer waiting after 
```
            await _concurrentAwaiter.WaitAsync().ConfigureAwait(false);
```
happens, its then "executing".

Not sure this really makes much difference though.

Reply:
You meant `Release()`. In my opinion release and output demand.